src/lib/ngc/compile-source-files.ts

```diff
@@ -159,11 +159,17 @@ export async function compileSourceFiles(
 
     // Only request Angular template diagnostics for affected files to avoid
     // overhead of template diagnostics for unchanged files.
-    if (affectedFiles.has(sourceFile)) {
-      const angularDiagnostics = angularCompiler.getDiagnosticsForFile(
-        sourceFile,
-        affectedFiles.size === 1 ? /** OptimizeFor.SingleFile **/ 0 : /** OptimizeFor.WholeProgram */ 1,
-      );
+    if (
+      affectedFiles.has(sourceFile) ||
+      !sourceFileCache.get(sourceFile.fileName)?.angularDiagnostics ||
+      sourceFileCache.get(sourceFile.fileName).angularDiagnostics.length
+    ) {
+      const angularDiagnostics =
+        (!affectedFiles.has(sourceFile) && sourceFileCache.get(sourceFile.fileName)?.angularDiagnostics) ||
+        angularCompiler.getDiagnosticsForFile(
+          sourceFile,
+          affectedFiles.size === 1 ? /** OptimizeFor.SingleFile **/ 0 : /** OptimizeFor.WholeProgram */ 1,
+        );
 
       allDiagnostics.push(...angularDiagnostics);
       sourceFileCache.updateAngularDiagnostics(sourceFile, angularDiagnostics);
```

Comment:
Consider the below, to improve readability

```suggestion
    if (affectedFiles.has(sourceFile)) {
      const angularDiagnostics = angularCompiler.getDiagnosticsForFile(
        sourceFile,
        affectedFiles.size === 1 ? /** OptimizeFor.SingleFile **/ 0 : /** OptimizeFor.WholeProgram */ 1,
      );

      allDiagnostics.push(...angularDiagnostics);
      sourceFileCache.updateAngularDiagnostics(sourceFile, angularDiagnostics);
    } else {
      const cachedDiagnostics = sourceFileCache.getAngularDiagnostics(sourceFile);
      if (cachedDiagnostics?.length) {
        allDiagnostics.push(...cachedDiagnostics);
      }
    }
```

Reply:
We still need to account for cases where the file in `FileCache` was deleted by the file watcher due to a change or indirect change (in case of an update to a component template file) and diagnostics were lost/need to be re-evaluated.

```ts
const cachedDiagnostics = sourceFileCache.getAngularDiagnostics(sourceFile);
if (affectedFiles.has(sourceFile) || !cachedDiagnostics) {
  ...
}
```

As I said before, sometimes the component file is not reported as affected after a few saves to its template file and that's why I also check if it exists in `FileCache`, and if not, I know it has changed and deleted by the watcher and I need to re-run diagnostics.
Maybe the fact that it is not reported as affected it's a bug in the compiler but I am not sure...

Reply:
Do you have any idea where the root cause of this problem is?

Reply:
Do you have a test case that I can test against?

Reply:
I tested you PR but I am still able to reproduce the problem. I am going to try and create the test cases.

Reply:
I have added a two tests that reproduce the issue.

The first one saves the same template file a few times, without making any change on it, expecting always an error but it ends up compiling successfully after a couple of saves.

In the second test that I added I tried to do multiple saves on a valid component file and then change it's input name expecting to get an error in a secondary entry point.

Reply:
thanks for this.

Reply:
My first approach on this issue was to force a new version of the source file for unchanged files. While there might be an issue with the angular compiler, would it be possible to do this with a workaround? I really don't know where the problem is or a better way to solve it.